clean_code/algorithms/creation_algorithms.py

```diff
@@ -0,0 +1,68 @@
+from math import factorial
+
+"""This module supplies fast algorithms for pattern-related creations:
+   all permutations of a staring, duplicate-deletion."""
+
+def create_trimmed_string(base_string):
+    """deletes duplicates in a string. example: aaabbaac -> abac"""
+    trimmed_elements = [base_string[0]]
+    for index in range(1, len(base_string)):
+        if base_string[index] != base_string[index-1]:
+            trimmed_elements.append(base_string[index])
+    return ''.join(trimmed_elements)
+
+
+def generate_permutations_iterably(elements):
+    """returns all the permutations of a string. the function is non-recursive"""
+    permutations = set()
+    final_permutations = set()
+    next_index_to_be_permutated = 0
+    expectd_permutations_number = factorial(len(elements))
+
+    new_permutations = _generate_new_permutations(elements)
+    permutations.update(new_permutations)
```

Comment:
The four names: `permutations`, `permutations_copy`, `new_permutations` and `final_permutations` is too confusing.